### PR TITLE
chore(ui): display API connector icons

### LINKED ISSUE #1029: API connector: uploaded image (icon) doesn't display anywhere
apiconnectoriconupload

### PR DESCRIPTION
**Pull request by zregvart**
If the icon is persisted in the database as opposed to statically served
from assets, point the icon url to the API endpoint for fetching the
icon.

Fixes #1029

### CODE REVIEW

**Comment by zregvart** [comment]
I'm sure that this can be done more elegantly, I just hammered at it until I got both the static and uploaded icons to show up. If no serious issues arise from this can we merge this for TP3 and refactor in TP4?

**Comment by zregvart** [comment]
Seems that it might have broken the new integration flow.

**Comment by zregvart** [comment]
False alarm, didn't have the Verifier running, so integration editor broke.

**Comment by gashcrumb** [comment]
> I'm sure that this can be done more elegantly, I just hammered at it until I got both the static and uploaded icons to show up. If no serious issues arise from this can we merge this for TP3 and refactor in TP4?

I dunno, there's some copy/paste in there and everything, are you gonna do this refactoring?

File: app/ui/src/app/connections/detail-page/info.component.ts [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
+import { Component, Input, Output, EventEmitter, SimpleChanges } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
 import { Observable } from 'rxjs/Observable';
 import { Subscription } from 'rxjs/Subscription';
 
 import { Connection } from '../../model';
 import { ConnectionService } from '../../store/connection/connection.service';
 import { ConnectionConfigurationService } from '../common/configuration/configuration.service';
+import { ConfigService } from '@syndesis/ui/config.service';
+import { OnChanges } from '@angular/core/src/metadata/lifecycle_hooks';
```
Comment (gashcrumb) [inline app/ui/src/app/connections/detail-page/info.component.ts]: You can just import this from `@angular/core`, no?

File: app/ui/src/app/connections/detail-page/info.component.ts [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
 
 @Component({
   selector: 'syndesis-connection-detail-info',
   template: `
     <h1>
       <dl class="dl-horizontal">
         <dt>
-          <img src="../../../assets/icons/{{ connection.connectorId }}.connection.png" height="46" width="46">
+          <img src="{{ icon }}" height="46" width="46">
```
Comment (gashcrumb) [inline app/ui/src/app/connections/detail-page/info.component.ts]: Shouldn't this be `[src]="icon"`?

Should really change that height/width to be set via stylesheet too, though I see the original code didn't bother to either.

File: app/ui/src/app/connections/detail-page/info.component.ts [hunk 3]
```diff
@@ -51,20 +53,33 @@ import { ConnectionConfigurationService } from '../common/configuration/configur
...
+  ) {
+    this.apiEndpoint = config.getSettings().apiEndpoint;
+  }
 
   onAttributeUpdated(attr: string, value) {
     this.connection[attr] = value;
     this.updated.emit(this.connection);
   }
 
+  ngOnChanges(changes: SimpleChanges) {
```
Comment (gashcrumb) [inline app/ui/src/app/connections/detail-page/info.component.ts]: This executes lots of times, and I suspect we don't need to evaluate this code that often.  This should probably go in `ngOnInit` where the `connection` object should also be available and set.

File: app/ui/src/app/connections/list/list.component.ts [hunk 4]
```diff
@@ -102,4 +107,10 @@ export class ConnectionsListComponent implements OnInit {
     return connection.id === this.selectedId;
   }
 
+  connectionIcon(connection: Connection) {
+    if (connection.icon.startsWith('db:')) {
+      return Observable.of(`${this.apiEndpoint}/connectors/${connection.connectorId || connection.id}/icon`);
```
Comment (gashcrumb) [inline app/ui/src/app/connections/list/list.component.ts]: Why is this async?  :-S

File: app/ui/src/app/customizations/api-connector/api-connector-detail/api-connector-detail.component.ts [hunk 5]
```diff
@@ -34,9 +40,17 @@ export class ApiConnectorDetailComponent implements OnInit {
       .switchMap(([id, apiConnectors]: [string, ApiConnectors]) =>
         apiConnectors.filter(apiConnector => apiConnector.id == id)
       );
+
+    this.icon$ = this.apiConnectorData$.mergeMap(connector => {
+      if (connector.icon.startsWith('db:')) {
```
Comment (gashcrumb) [inline app/ui/src/app/customizations/api-connector/api-connector-detail/api-connector-detail.component.ts]: Have seen this `if` statement a lot.  Maybe there's a better place to check this.  My suggestion would be to enhance the service and maybe just reset/add the icon property in there, as with this current code the person refactoring it has to go through all these files and discover/update these if statements.

File: app/ui/src/app/dashboard/connections.component.html [hunk 6]
```diff
@@ -32,7 +32,7 @@ <h1>Create a Connection</h1>
 
               <div class="card-pf-top-element">
                 <span class="card-pf-icon-large image-icon">
-                <img src="../../../assets/icons/{{ connection.connectorId }}.connection.png">
+                <img src="{{ connectionIcon(connection) }}">
```
Comment (gashcrumb) [inline app/ui/src/app/dashboard/connections.component.html]: Yeah, definitely work looking into updating the store.  Like we could add an override to this class:

https://github.com/syndesisio/syndesis/blob/master/app/ui/src/app/store/connection/connection.service.ts

That basically sets the icon, you'd override this function:

https://github.com/syndesisio/syndesis/blob/master/app/ui/src/app/store/entity/entity.store.ts#L60-L75

and probably modify this too:

https://github.com/syndesisio/syndesis/blob/master/app/ui/src/app/store/entity/entity.store.ts#L77-L86

to update/enhance the `icon` property on the connection/connector.  Probably need to do it for both the connection and the connector stores.

File: app/ui/src/app/dashboard/connections.component.ts [hunk 7]
```diff
@@ -28,6 +30,12 @@ export class DashboardConnectionsComponent implements OnInit {
 
   truncateTrail = '…';
 
+  constructor(
+    private config: ConfigService
```
Comment (gashcrumb) [inline app/ui/src/app/dashboard/connections.component.ts]: Yeah, if we put this in the store then you don't need all these `ConfigService` injections etc.

**Comment by zregvart** [comment]
Impressive work @deeleman 🥇

**Pull request merged**